File: java/command_injection/SystemUtils.java
import java.io.*;

public class SystemUtils {
    // VULN 1: Runtime.exec with string concatenation - ping
    public boolean checkHostAvailability(String hostname) throws IOException {
        Process process = Runtime.getRuntime().exec("ping -c 4 " + hostname);
        try {
            return process.waitFor() == 0;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return false;
        }
    }

    // VULN 2: ProcessBuilder sh -c with string concatenation - traceroute
    public String traceNetworkRoute(String destination) throws IOException {
        ProcessBuilder pb = new ProcessBuilder("sh", "-c", "traceroute " + destination);
        Process process = pb.start();
        return new String(process.getInputStream().readAllBytes());
    }

    // VULN 3: Runtime.exec with string concatenation - nslookup
    public String resolveHostname(String hostname) throws IOException {
        Process process = Runtime.getRuntime().exec("nslookup " + hostname);
        return new String(process.getInputStream().readAllBytes());
    }
}

File: dotnet/command_injection/SystemUtils.cs
using System.Diagnostics;

public class SystemUtils
{
    // VULN 1: ProcessStartInfo with string concatenation - ping
    public bool CheckHostAvailability(string hostname)
    {
        var psi = new ProcessStartInfo("ping", "-c 4 " + hostname)
        {
            RedirectStandardOutput = true,
            UseShellExecute = false,
        };
        var process = Process.Start(psi);
        process.WaitForExit();
        return process.ExitCode == 0;
    }

    // VULN 2: ProcessStartInfo sh -c with string interpolation - traceroute
    public string TraceNetworkRoute(string destination)
    {
        var psi = new ProcessStartInfo("sh", $"-c \"traceroute {destination}\"")
        {
            RedirectStandardOutput = true,
            UseShellExecute = false,
        };
        var process = Process.Start(psi);
        return process.StandardOutput.ReadToEnd();
    }

    // VULN 3: ProcessStartInfo sh -c with string concatenation - nslookup
    public string ResolveHostname(string hostname)
    {
        var psi = new ProcessStartInfo("sh", "-c nslookup " + hostname)
        {
            RedirectStandardOutput = true,
            UseShellExecute = false,
        };
        var process = Process.Start(psi);
        return process.StandardOutput.ReadToEnd();
    }
}
